File: src/schemas/terminal.ts
import { z } from "zod"

export const terminalKindSchema = z.enum(["bash", "codex", "claude"])
export const terminalStatusSchema = z.enum(["running", "exited", "failed"])

export const terminalSessionSchema = z
  .object({
    id: z.string().uuid(),
    kind: terminalKindSchema,
    title: z.string().min(1),
    status: terminalStatusSchema,
    workingDirectory: z.string(),
    createdAt: z.string().datetime({ offset: true }),
  })
  .strict()

export const terminalOutputEventSchema = z
  .object({
    sessionId: z.string().uuid(),
    data: z.string(),
  })
  .strict()

export const terminalExitEventSchema = z
  .object({
    sessionId: z.string().uuid(),
    status: terminalStatusSchema,
    exitCode: z.number().int().nonnegative().nullable().optional(),
  })
  .strict()

export const createTerminalSessionInputSchema = z
  .object({
    kind: terminalKindSchema,
    rows: z.number().int().min(2).max(500),
    cols: z.number().int().min(2).max(500),
  })
  .strict()

export const writeTerminalSessionInputSchema = z
  .object({
    sessionId: z.string().uuid(),
    data: z.string().max(64 * 1024),
  })
  .strict()

export const resizeTerminalSessionInputSchema = z
  .object({
    sessionId: z.string().uuid(),
    rows: z.number().int().min(2).max(500),
    cols: z.number().int().min(2).max(500),
  })
  .strict()

export const closeTerminalSessionInputSchema = z
  .object({
    sessionId: z.string().uuid(),
  })
  .strict()

export type TerminalKind = z.infer<typeof terminalKindSchema>
export type TerminalSession = z.infer<typeof terminalSessionSchema>

File: src-tauri/src/domain/terminal.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "lowercase")]
pub enum TerminalKind {
    Bash,
    Codex,
    Claude,
}

impl TerminalKind {
    pub const fn title(self) -> &'static str {
        match self {
            Self::Bash => "Shell",
            Self::Codex => "Codex",
            Self::Claude => "Claude",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum TerminalStatus {
    Running,
    Exited,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct TerminalSession {
    pub id: Uuid,
    pub kind: TerminalKind,
    pub title: String,
    pub status: TerminalStatus,
    pub working_directory: String,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct TerminalOutputEvent {
    pub session_id: Uuid,
    pub data: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct TerminalExitEvent {
    pub session_id: Uuid,
    pub status: TerminalStatus,
    pub exit_code: Option<u32>,
}
